src/main/java/org/wildfly/security/password/spec/DefaultPasswordSpecEncoding.java

```diff
@@ -0,0 +1,233 @@
+     */
+    public static PasswordSpec decode(byte[] encoded) {
+        ByteBuffer buffer = ByteBuffer.wrap(encoded);
+        byte identifier;
+
+        try {
+            identifier = buffer.get();
+        } catch (Exception e) {
+            throw ElytronMessages.log.couldNotObtainKeySpecEncodingIdentifier();
+        }
+
+        switch (identifier) {
+            case CLEAR_PASSWORD_SPEC_ID:
+                return decodeClearPasswordSpec(buffer);
+            case DIGEST_PASSWORD_SPEC_ID:
+                return decodeDigestPasswordSpec(buffer);
+            case HASH_PASSWORD_SPEC_ID:
+                return decodeHashPasswordSpec(buffer);
+            case SALTED_HASH_PASSWORD_SPEC_ID:
+                return decodeSaltedHashPasswordSpec(buffer);
+            case ITERATED_SALTED_HASH_SPEC_ID:
+                return decodeIteratedSaltedHashPasswordSpec(buffer);
+            default:
+                return null;
+        }
+    }
+
+    private static byte[] encodeIteratedSaltedHashSpec(Password password, PasswordFactory passwordFactory) throws InvalidKeySpecException {
+        IteratedSaltedHashPasswordSpec keySpec = passwordFactory.getKeySpec(password, IteratedSaltedHashPasswordSpec.class);
+        byte[] hash1 = keySpec.getHash();
+        ByteBuffer specBytes = ByteBuffer.allocate(1 + 4 + SALT_SIZE + hash1.length)
+                .put(ITERATED_SALTED_HASH_SPEC_ID) // 1 byte identifying the password spec
+                .putInt(keySpec.getIterationCount()); // 4 bytes representing the iteration count
+        // SALT_SIZE bytes representing the salt, 0s are appended to the initial bytes to fill unused bytes.
+        byte[] salt = keySpec.getSalt();
+        encodeSalt(specBytes, salt);
+        specBytes.put(salt);
+        specBytes.put(hash1); // the remaining bytes representing the hash
+        return specBytes.array();
+    }
+
+    private static PasswordSpec decodeIteratedSaltedHashPasswordSpec(ByteBuffer buffer) {
+        // decode the iteration count
+        int iterationCount1 = buffer.getInt();
+        // decode the rawSalt
+        byte[] rawSalt = new byte[SALT_SIZE];
+        buffer.get(rawSalt);
+        byte[] salt = decodeSalt(rawSalt);
+        // decode the hash
+        byte[] hash1 = new byte[buffer.remaining()];
+        buffer.get(hash1);
+        return new IteratedSaltedHashPasswordSpec(hash1, salt, iterationCount1);
+    }
+
+    private static byte[] encodeSaltedHashPasswordSpec(Password password, PasswordFactory passwordFactory) throws InvalidKeySpecException {
+        SaltedHashPasswordSpec keySpec = passwordFactory.getKeySpec(password, SaltedHashPasswordSpec.class);
+        byte[] hash = keySpec.getHash();
+        ByteBuffer specBytes = ByteBuffer.allocate(1 + SALT_SIZE + hash.length)
+                .put(SALTED_HASH_PASSWORD_SPEC_ID); // 1 byte identifying the password spec
+        // SALT_SIZE bytes representing the salt, 0s are appended to the initial bytes to fill unused bytes.
```

Comment:
Using fixed-length fields does provide a solution here, but two other options exist: pack the entire field as a packed int (wastes 1 bit per byte), or provide a packed-int length followed by raw bytes (more efficient for longer fields).

Here's the breakdown of spatial efficiency:

Size | Length Field | Packed
---- | ------------ | ------
0 | 1 | 0
1 | 2 | 2
2 | 3 | 3
3 | 4 | 4
4 | 5 | 5
5 | 6 | 6
6 | 7 | 7
7 | 8 | 8
8 | 9 | 10
9 | 10 | 11
10 | 11 | 12
11 | 12 | 13
12 | 13 | 14
13 | 14 | 15
14 | 15 | 16
15 | 16 | 18
16 | 17 | 19
17 | 18 | 20
18 | 19 | 21
19 | 20 | 22
20 | 21 | 23
21 | 22 | 24
22 | 23 | 26
23 | 24 | 27
24 | 25 | 28
25 | 26 | 29
26 | 27 | 30
27 | 28 | 31

So packed is better when the field is up to 7 bytes long, after which a packed length field + raw bytes is better.